libnd4j/include/ops/declarable/generic/nn/dot_product_attention.cpp

```diff
@@ -0,0 +1,223 @@
+#include <ops/declarable/helpers/reverse.h>
+
+
+namespace nd4j {
+namespace ops  {
+
+    CUSTOM_OP_IMPL(dot_product_attention, 3, -1, false, 0, 2) {
+        auto queries = INPUT_VARIABLE(0);
+        auto keys = INPUT_VARIABLE(1);
+        auto values = INPUT_VARIABLE(2);
+        auto mask    = block.width() > 3 ? INPUT_VARIABLE(3) : nullptr;
+
+        auto output = OUTPUT_VARIABLE(0);
+        NDArray* weights;
+        bool outputWeights = INT_ARG(1);
+        if(outputWeights){
+            weights = OUTPUT_VARIABLE(1);
+        }else{
+            auto weightShape = ShapeUtils::evalShapeForMatmul(keys->getShapeInfo(), queries->getShapeInfo(), true, false);
+            weights = NDArrayFactory::create_('c', weightShape, values->dataType(), block.workspace());
+        }
+
+        int normalization = INT_ARG(0);
+
+        REQUIRE_TRUE(queries->rankOf() == keys->rankOf() && keys->rankOf() == values->rankOf(), 0,
+                     "dot_product_attention: Queries, Keys and Values must have same rank. "
+                     "But got queries = %s, keys = %s, values = %s", ShapeUtils::shapeAsString(queries).c_str(),
+                     ShapeUtils::shapeAsString(keys).c_str(), ShapeUtils::shapeAsString(values).c_str());
+
+        REQUIRE_TRUE(queries->rankOf() == 3 || queries->rankOf() == 4, 0,
+                     "dot_product_attention: Queries, Keys and Values must be rank 3 arrays for single headed attention "
+                     "or rank 4 arrays for multi headed attention. But got rank = %i", queries->rankOf());
+
+        REQUIRE_TRUE(queries->sizeAt(0) == keys->sizeAt(0) && keys->sizeAt(0) == values->sizeAt(0), 0,
+                "dot_product_attention: Queries, Keys and Values must have the same mini batch size. "
+                "But got queries = %i, keys = %i, values = %i", queries->sizeAt(0), keys->sizeAt(0), values->sizeAt(0));
+
+        REQUIRE_TRUE(queries->sizeAt(-2) == keys->sizeAt(-2), 0,
+                "dot_product_attention: Queries and Keys must have the same feature size. "
+                "But got queries = %i, keys = %i", queries->sizeAt(-2), keys->sizeAt(-2));
+
+        REQUIRE_TRUE(keys->sizeAt(-1) == values->sizeAt(-1), 0,
+                "dot_product_attention: Keys and Values must have the same timestep length. "
+                "But got keys = %i, values = %i", keys->sizeAt(-1), values->sizeAt(-1));
+
+        nd4j::ops::matmul mmul;
+        mmul.execute({keys, queries}, {weights}, {}, {1}, {});
+        if(normalization) {
+            *weights /= sqrt((double)keys->sizeAt(-2));
+        }
+
+        if(mask != nullptr){
+            NDArray* reshapedMask;
+            if(weights->rankOf() == 4){
+                reshapedMask = mask->reshape(mask->ordering(), {mask->sizeAt(0), 1, mask->sizeAt(1), 1});
+            }else{
+                reshapedMask = mask->reshape(mask->ordering(), {mask->sizeAt(0), mask->sizeAt(1), 1});
+            }
+            *weights += (*reshapedMask - 1) * 1e9;
```

Comment:
Maybe add a comment here - seemed a little strange on first glance.
I gather you want masked steps to have very large negative attention weights as a way of doing "masked softmax"?
How was the 1e9 chosen? (just wondering if we'll even have numerical stability issues...)

Reply:
The 1e9 is used by the tensor2tensor package as well. Bert on the other hand uses just 1e4, while GPT-2 uses 1e10.